docs/design/values.md

```diff
@@ -0,0 +1,973 @@
+defined to create a new local copy of the argument, although it may move into
+this copy.
+
+Carbon's values are much closer to a `const &` in C++ with extra restrictions
+such as allowing copies under "as-if" and preventing taking the address.
+Combined, these restrictions allow implementation strategies such as in-register
+parameters.
+
+### Value representation and customization
+
+The representation of a value expression is especially important because it
+forms the calling convention used for the vast majority of function parameters
+-- function inputs. Given this importance, it's important that it is predictable
+and customizable by the value's type. Similarly, while Carbon code must be
+correct with either a copy or a reference-based implementation, we want which
+implementation strategy is used to be a predictable and customizable property of
+the type of a value.
+
+A type can optionally control its value representation using a custom syntax
+similar to customizing its [destructor](/docs/design/classes.md#destructors).
+This syntax sets the representation to some type uses a keyword `value_rep` and
+can appear where a member declaration would be valid within the type:
+
+```carbon
+class SomeType {
+  value_rep = RepresentationType;
+}
+```
+
+**Open question:** The syntax for this is just placeholder, using a placeholder
+keyword. It isn't final at all and likely will need to change to read well.
+
+The provided representation type must be one of the following:
+
+-   `const Self` -- this forces the use of a _copy_ of the object.
+-   `const Self *` -- this forces the use of a [_pointer_](#pointers) to the
+    original object.
+-   A custom type that is not `Self`.
+
+If the representation is `const Self` or `const Self *`, then the type fields
+will be accessible as [_value expressions_](#value-expressions) using the normal
+member access syntax for value expressions of a type. These will be implemented
+by either accessing a copy of the object in the non-pointer case or a pointer to
+the original object in the pointer case. A representation of `const Self`
+requires copying to be valid for the type. This provides the builtin
+functionality but allows explicitly controlling which representation should be
+used.
+
+If no customization is provided, the implementation will select one based on a
+set of heuristics. Some examples:
+
+-   Non-copyable types and polymorphic types would use a `const Self*`.
+-   Small objects that are trivially copied in a machine register would use
+    `const Self`.
+
+When a custom type is provided, it must not be `Self` (or the other two cases).
+The type provided will be used on function call boundaries and as the
+implementation representation for `let` bindings and other value expressions
+referencing an object of the type. A specifier of `value_rep = T;` will require
+that the type `impls ReferenceImplicitAs(T)` using the following interface:
+
+```carbon
+interface ReferenceImplicitAs(T:! type) {
```

Comment:
From discussion:
- should make `T` an associated type.
- could make `value_rep = T;` act as a forward declaration of `impl as ReferenceImplicitAs where .Dest = T`; this would mean that the definition could just write `impl as ReferenceImplicitAs where _`